thunderbirdparser/src/org/sleuthkit/autopsy/thunderbirdparser/EmailMessageThreader.java

```diff
@@ -0,0 +1,674 @@
+/*
+ * Autopsy Forensic Browser
+ *
+ * Copyright 2019 Basis Technology Corp.
+ * Contact: carrier <at> sleuthkit <dot> org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.sleuthkit.autopsy.thunderbirdparser;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Given a list of email messages arranges the message into threads using the
+ * message reference lists.
+ *
+ * This threader is based heavely off of the algorithum found at
+ * <a href="https://www.jwz.org/doc/threading.html">
+ * "message threading." by Jamie Zawinski</a>
+ *
+ */
+final class EmailMessageThreader {
+
+    private int bogus_id_count = 0;
+
+    public void threadMessages(List<EmailMessage> emailMessages, String threadIDPrefix) {
+        HashMap<String, Container> id_table = createIDTable(emailMessages);
+        Set<Container> rootSet = getRootSet(id_table);
+
+        pruneEmptyContainers(rootSet);
+
+        Set<Container> finalSet = groupBySubject(rootSet);
+
+        assignThreadIDs(finalSet, threadIDPrefix);
+    }
+
+    /**
+     * Walks the list of emailMessages creating a Container object for each
+     * unique message ID found. Adds the emailMessage to the container where
+     * possible.
+     *
+     * @param emailMessages
+     *
+     * @return - HashMap of all message where the key is the message-ID of the message
+     */
+    private HashMap<String, Container> createIDTable(List<EmailMessage> emailMessages) {
+        HashMap<String, Container> id_table = new HashMap<>();
+
+        for (EmailMessage message : emailMessages) {
+            String messageID = message.getMessageID();
+
+            // Check the id_table for an existing Container for message-id
+            Container container = id_table.get(messageID);
+
+            // An existing container for message-id was found
+            if (container != null) {
+                // If the existing Container has a message already assocated with it
+                // emailMessage maybe a duplicate, so we don't lose the existance of 
+                // the duplicate message assign it a bogus message-id
+                if (container.hasMessage()) {
```

Comment:
In what scenario would we have a Container in the map without a message?

Reply:
Just saw it from below.  We should add a line here that it could be empty if we've seen a reference to this ID, but haven't seen the actual message yet.

Reply:
When the initial list of message ids is created we go through the message and then its list of reference's.  When we process the reference list we only have the ID, not the actual message (its assumed that we will probably come across it later).  If we don't ever come across that message, the container would exist without the message.